Automattic/wp-calypso
# Happychat: Refactor out helper.js for clearer loading states

This PR lays foundation for improving Happychat connection statuses to handle lost connections and reconnects (https://github.com/Automattic/wp-calypso/issues/12183), by refactoring out the `helper.js` which primarily existed to show interface elements in various states. Most of this work comes from @jordwest's PR https://github.com/Automattic/wp-calypso/pull/12219, which I'm splitting up into two parts — the next PR will introduce new connection statuses for various connectivity issues.

The screenshots below compare the current Happychat load experience with the new one. This shows on the `/me/chat` page where HC is visible from the beginning, but in practice most users will interact through the sidebar HC. That sidebar usually pre-connects before it's visible, so most of the time users won't see these loading states. But simplifying in this way will help with the following PR, where we'll give users more feedback about connection issues.

## Screenshot comparison

#### Current experience

_Uninitialized:_
![screen shot 2017-03-31 at 2 52 20 pm](https://cloud.githubusercontent.com/assets/518059/24567066/baf62f68-1621-11e7-926c-7d158c2ad91f.png)

_Connecting:_
![screen shot 2017-03-31 at 2 52 25 pm](https://cloud.githubusercontent.com/assets/518059/24567067/baf79164-1621-11e7-9554-6693cef2a98f.png)

_Connected:_
![screen shot 2017-03-31 at 2 52 31 pm](https://cloud.githubusercontent.com/assets/518059/24567065/baf41a84-1621-11e7-950f-5d8d92ece3b7.png)

#### New experience

_Uninitialized:_
![screen shot 2017-03-31 at 2 45 42 pm](https://cloud.githubusercontent.com/assets/518059/24567075/c94b8194-1621-11e7-996c-acc509640fee.png)

_Connecting:_
![screen shot 2017-03-31 at 2 45 53 pm](https://cloud.githubusercontent.com/assets/518059/24567076/c94bd752-1621-11e7-95fe-b99974155ccd.png)

_Connected:_
![screen shot 2017-03-31 at 2 45 59 pm](https://cloud.githubusercontent.com/assets/518059/24567074/c94ad4a6-1621-11e7-8b23-99ec09fa53e2.png)

## To test

Make sure that normal paths to Happychat communication work as expected. You can use Redux Dev Tools to scrub through the various loading states.
